### DISCUSSION TITLE
How to write a custom data type with multiple values

### DISCUSSION THREAD

**Discussion by stfnndrsn** [FreeOpcUa/opcua-asyncio#1278] [Q&A] [answered]
Hello there! 

I'm trying to write to a node that hos a custom struct datatype; but I can't figure out how to write the object. 

Example of the datatype/input from a OPC viewer:

![image](https://user-images.githubusercontent.com/28970/231179926-71b4b8ce-d3bc-48a7-8f06-6522ccab6f6f.png)

I've tried several of ways to make the object and write it to the node; but I have not got it right yet.

```python
    async with client:
        _logger.warning("Connected")
        type_defs  = await client.load_type_definitions()  # scan server for custom structures and import them

        await client.load_type_definitions()
        node = client.get_node('ns=3;s="DB_Com_       "."In"')

        _logger.info("Node: %r", node)

        # Use the custom struct
        answer = ua._DB_Com_In_(
            ID=99, 
            Value1=100, 
            Value2=199
        )

        # What about a Dict?
        answer = {
            "ID": ua.DataValue(ua.Variant(39, ua.VariantType.UInt32)), 
            "Value1": ua.DataValue(ua.Variant(100, ua.VariantType.UInt32)), 
            "Value2": ua.DataValue(ua.Variant(199, ua.VariantType.UInt32))
        }

        # Struct with DataValue and types
        answer = ua._DB_Com_Hobbii_In_(
            ID=ua.DataValue(ua.Variant(39, ua.VariantType.UInt32)), 
            Value1=ua.DataValue(ua.Variant(100, ua.VariantType.UInt32)), 
            Value2=ua.DataValue(ua.Va199t(9102, ua.VariantType.UInt32))
        )

        await node.write_value(answer)

``` 

Any pointer into the right direction will be appreciated :D

**Comment by schroeder-** [comment]
Your problem is the name of the datatype. We have a special converter for the name, but i can't tell you what the name of the struct is.
The best way is to set the log to info/debug and search the log output for the added datatype. Or check type_defs  for the struct name. 
Also maybe the server doesn't support `load_type_definitions `and you could try `load_data_type_definitions `instead.
For constructing a custom datatype this should work:
```Python
answer = ua.DB_Com_In_(
            ID=99, 
            Destination1=100, 
            Destination2=199
        )
```

**Comment by stfnndrsn** [comment]
Thanks @schroeder- 

Sorry for obfuscation the identifier a bit; but the server seems to support both `load_type_definitions` and `load_data_type_definitions`.

```
INFO:asyncua.common.structures104:Registering data type NodeId(Identifier='DT_"DB_Com_xx"."In"', NamespaceIndex=3, NodeIdType=<NodeIdType.String: 3>) QualifiedName(NamespaceIndex=3, Name='"DB_Com_xx"."In"')
WARNING:asyncua.common.structures104:renamed "DB_Com_xx"."In" to _DB_Com_xx_In_ due to Python syntax
```
Result from the write operation:

```
asyncua.ua.uaerrors._auto.BadWriteNotSupported: "The server does not support writing the combination of value, status and timestamps provided."(BadWriteNotSupported)
```

**Reply by schroeder-** [reply]
You try to write the structure definition in your example. You must write a node of the nodetype variable. Best use the nodeid you used for the UaExpert screenshot.

**Reply by stfnndrsn** [reply]
Hmm.. The naming for the type and nodeid is pretty similar

Node:
```
INFO:__main__:Node: Node(NodeId(Identifier='"DB_Com_xx"."In"', NamespaceIndex=3, NodeIdType=<NodeIdType.String: 3>)) 
```

Type (value as read from PLC from the node):
```
_DB_Com_xx_In_(Value1=9101, Value2=9102, ID=40)
```

Or I'm probably missing something important. :/

**Reply by schroeder-** [reply, marked as answer]
Maybe the server doens't like our DataValue with SourceTimestamp.
```Python
answer = ua._DB_Com_In_(
            ID=99, 
            Value1=100, 
            Value2=199
        )
  await node.write_value(ua.DataValue(answer))
```

**Reply by stfnndrsn** [reply]
Thanks :)

It seems to be accepted by the PLC / OPC UA Server.
